enhancements/installer/storage-class-encrypted.md

```diff
@@ -0,0 +1,278 @@
+---
+title: encrypted-default-storage-class
+authors:
+  - @patrickdillon
+reviewers: # Include a comment about what domain expertise a reviewer is expected to bring and what area of the enhancement you expect them to focus on. For example: - "@networkguru, for networking aspects, please look at IP bootstrapping aspect"
+  - @jhixson74
+  - @jsafrane
+  - @bertinatto
+  - @trilokgeer
+approvers:
+  - TBD
+api-approvers: # In case of new or modified APIs or API extensions (CRDs, aggregated apiservers, webhooks, finalizers). If there is no API change, use "None"
+  - TBD
+creation-date: 2022-06-22
+last-updated: 2022-06-22
+tracking-link: # link to the tracking ticket (for example: Jira Feature or Epic ticket) that corresponds to this enhancement
+  - [CORS-2049](https://issues.redhat.com/browse/CORS-2049)
+  - [OCPPLAN-9307](https://issues.redhat.com/browse/OCPPLAN-9307)
+---
+
+# Default Storage Class with Encrypted Keys 
+
+## Summary
+
+When user-managed encryption keys are provided at install time, the default storage
+class should be set to utilize those keys. Presently, only root volumes for nodes are
+encrypted with those keys--but users would expect the persistent storage volumes to
+be encrypted as well. This enhancement proposes a solution for AWS, GCP, & Azure.
+
+## Motivation
+
+The primary motivation is to enable managed services to encrypt storage volumes with
+customer-provided keys automatically as part of the installation process without requiring
+day-2 operations. The behavior of automatically setting the default storage class is
+sensible for individual (non-managed-service) users as well; setting the default will
+reduce their day-2 burden and decrease the opportunity for user error.
+
+### User Stories
+
+When I install a cluster, I want to provide user-managed encryption keys in the install-config
+so that the default storage class will be set to provision volumes with those keys.
+
+### Goals
+
+* Default storage class is set to use user-provided keys at cluster creation
+
+### Non-Goals
+
+
+## Proposal
+
+### Workflow Description
+
+1. Cluster creator specifies (aws|gcp|azure) user-managed encryption keys 
+in `defaultMachinePlatform` of install config.
+2. Installer adds encryption key fields to cluster infrastructure object.
+3. Cluster-storage-operator/CSI drivers use cluster infrastructure object
+to create default storage class using user-encrypted keys from infra object.
+
+
+### API Extensions
+
+This enhancement extends the cluster `Infrastructure` API object to add fields
+for the user-managed encryption keys. The fields will be added to the platform
+spec and status for each platform. They are added to the spec so that users can edit the
+spec day 2 in order to update the default storage class. The status field would
+be updated by the cluster-storage operator/CSI driver.
```

Comment:
Why doesn't the installer set the value on the default storage class directly?

What happens if the user modifies the default storage class definition but does not update the Infrastructure API? Which settings are used?

Reply:
To add to Jan's answer, anything that can be handled outside of the installer should be.

> What happens if the user modifies the default storage class definition but does not update the Infrastructure API? Which settings are used?

I do not have much experience writing operators, but don't we have an established pattern for handling this problem in general? In this case, I would expect the CSI driver to overwrite the new user-specified class with the one specified in the infra object. But I would trust those with more operator experience here.

Reply:
> > Why doesn't the installer set the value on the default storage class directly?
> 
> StorageClasses are created and reconciled by CSI driver operators and not by installer. And the driver operator should update the StorageClass with new / deleted key ID. Cluster admins always have option to create a different StorageClasses with different keys and with #1152 they can update (or remove) the StorageClasses created by our operators.

Is the CSI driver operator already reconciling the config API to manage the default storage class settings, or is that net new behavior?

Reply:
Yes, it is reconciling SC parameters. For example, we changed the default volume type from pd-standard to pd-ssd in 4.11.